### PR TITLE
Upgrade ruff

### PR DESCRIPTION
**Pull request by Lee-W**
## Description

* relax ruff and black constraint
* upgrade dev depedencies
* use `ruff format`

closes: https://github.com/commitizen-tools/commitizen/pull/915

## Checklist

- [ ] Add test cases to all the changes you introduce
- [ ] Run `./scripts/format` and `./scripts/test` locally to ensure this change passes linter check and test
- [ ] Test the changes on the local machine manually
- [ ] Update the documentation for the changes

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -58,9 +59,9 @@ pytest-regressions = "^2.4.0"
 pytest-freezer = "^0.4.6"
 pytest-xdist = "^3.1.0"
 # code formatter
-black = ">=22.10,<24.0"
+black = ">=22.10"
```
[the remark this answers is not in the document]
Comment (Lee-W) [reply pyproject.toml:62]: good point! let me update it to black~=23.11 on both sides

**Review by noirbizarre — APPROVED** [review]
Note: I have been deploying this on my repositories lately: https://github.com/GabDug/sync-pre-commit-lock
It's a game changer: it update pre-commit hooks and keep them in sync when locking using `poetry` (I use it with `pdm` but same behavior)

**Comment by Lee-W** [comment]
> Note: I have been deploying this on my repositories lately: https://github.com/GabDug/sync-pre-commit-lock It's a game changer: it update pre-commit hooks and keep them in sync when locking using `poetry` (I use it with `pdm` but same behavior)

This indeed looks cool! How do you feel about `pdm`? I was thinking of trying it in my personal project as well

**Comment by Lee-W** [comment]
> I am a `pdm` contributor and maintainer. I use it on all my personal projects (or at least I am migrating them) and I am currently migrating my clients to `pdm` too. So far it's just so good !!
> 
> Note that I am currently working with a `copier`+ `pdm` + `commitizen` + pre-commit` stack, it's just perfect 👌🏼

Sounds cool! Will try it in my next project and see how it works 👀

**Pull request merged**